### DISCUSSION TITLE
Fish hook-env on cd is disabled after prompt render

### DISCUSSION THREAD

**Discussion by syhol** [jdx/mise#3886] [General]
### Summary

I just discovered that "hook-env on cd" is disabled as soon as an interactive prompt is rendered on fish, but not bash or zsh.

Feels like that should be consistent. I don't mind which way we go with it, we could even make it configurable with (yet another) activation flag

If we do it bash/zsh style and never disable hook-env on cd, then the case for shims is greatly reduced.

### Recreating 

<details>
<summary>First setup</summary>

#### First setup
```console
❯ docker run --pull=always -it --rm --entrypoint bash jdxcode/mise:latest
latest: Pulling from jdxcode/mise
Digest: sha256:ff2890f80d8352ae4424a03a5feca8e4a82f01e8e0181ca6287f40201d73fa6a
Status: Image is up to date for jdxcode/mise:latest
root@39d620693a55:/mise# apt update; apt install fish zsh;
...
Setting up zsh (5.9-4+b5) ...
Setting up fish (3.6.0-3.1+deb12u1) ...
...
root@39d620693a55:/mise# echo $PATH
/mise/shims:/usr/local/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@39d620693a55:/mise# export PATH="$(echo $PATH | sed 's#/mise/shims:##g')"
root@39d620693a55:/mise# echo $PATH
/usr/local/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@39d620693a55:/mise# mkdir -p /srv/proj1 /srv/proj2
root@39d620693a55:/mise# mise use node@20 --cd /srv/proj1
gpg: Signature made Wed Nov 20 14:34:05 2024 UTC
gpg:                using RSA key CC68F5A3106FF448322E48ED27F5E38D5B0A215F
gpg: Good signature from "marco-ippolito <marcoippolito54@gmail.com>" [unknown]
mise node@20.18.1 ✓ installed
mise /srv/proj1/mise.toml tools: node@20.18.1
root@39d620693a55:/mise# mise use node@22 --cd /srv/proj2
gpg: Signature made Tue Dec  3 20:37:27 2024 UTC
gpg:                using RSA key 108F52B48DB57BB0CC439B2997B01419BD92F80A
gpg: Good signature from "Ruy Adorno <ruyadorno@hotmail.com>" [unknown]
mise node@22.12.0 ✓ installed
mise /srv/proj2/mise.toml tools: node@22.12.0
```

</details>

<details>
<summary>Bash</summary>

#### Bash - `activate` and `cd` in the same command

```console
root@39d620693a55:/mise# mise deactivate 2>/dev/null
root@39d620693a55:/mise# eval "$(mise activate bash)" && cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
v20.18.1
v22.12.0
root@39d620693a55:/mise# mise deactivate
```

#### Bash - `activate` and `cd` in the different commands

```console
root@39d620693a55:/mise# mise deactivate 2>/dev/null
root@39d620693a55:/mise# eval "$(mise activate bash)"
root@39d620693a55:/mise# cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
v20.18.1
v22.12.0
root@39d620693a55:/mise# mise deactivate
```

</details>

<details>
<summary>Zsh</summary>

#### Into zsh

```console
root@39d620693a55:/mise# zsh
39d620693a55#
```

#### Zsh - `activate` and `cd` in the same command

```console
39d620693a55# mise deactivate 2>/dev/null
39d620693a55# eval "$(mise activate zsh)" && cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
v20.18.1
v22.12.0
39d620693a55# mise deactivate
```

#### Zsh - `activate` and `cd` in the different commands

```console
39d620693a55# mise deactivate 2>/dev/null
39d620693a55# eval "$(mise activate zsh)"
39d620693a55# cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
v20.18.1
v22.12.0
39d620693a55# mise deactivate
```

</details>

#### Into fish

```console
39d620693a55# exit
root@39d620693a55:/mise# fish
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
root@39d620693a55 /mise#
```

#### Fish - `activate` and `cd` in the same command

```console
root@39d620693a55 /mise# mise deactivate 2>/dev/null
root@39d620693a55 /mise# eval "$(mise activate fish)" && cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
v20.18.1
v22.12.0
root@39d620693a55 /mise# mise deactivate
```

#### Fish - `activate` and `cd` in the different commands

> [!CAUTION]
> This behaviour is different from the other shells where it works

```console
root@39d620693a55 /mise# mise deactivate 2>/dev/null
root@39d620693a55 /mise# eval "$(mise activate fish)"
root@39d620693a55 /mise# cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
fish: Unknown command: node
fish:
cd /srv/proj1 && node -v && cd /srv/proj2 && node -v && cd /mise
                 ^~~^
root@39d620693a55 /s/proj1 [127]# mise deactivate
```

Thats triggered by this: https://github.com/jdx/mise/blob/main/src/shell/fish.rs#L79

**Comment by jdx** [comment]
I believe I copied that logic from direnv which prevents running `hook-env` multiple times if you aren't changing the directory. It might need to be removed for us since.

**Comment by Marukome0743** [comment]
Thanks for the detailed report. This has now been fixed by [#11478](https://github.com/jdx/mise/pull/11478), which has been merged.

In the default Fish activation mode, `fish_preexec` no longer removes the PWD hook. This means that even when `mise activate fish` is evaluated as a separate command, subsequent commands such as `cd /srv/proj1 && node -v && cd /srv/proj2 && node -v` update the mise environment immediately after each directory change, matching the Bash and Zsh behavior.

The existing delayed behavior for `eval_after_arrow` and the prompt-only behavior for `disable_arrow` remain unchanged. Regression coverage was also added for consecutive directory changes in Fish.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
